Added new slots for the spright-chat-message to allow adding action buttons and follow-up prompts (#2552)
ni/nimble: packages/spright-components/src/chat/specs/README.md

#### Chat message

The `spright-chat-message` has the following slot elements.

> 1. `default` slot displays arbitrary slotted content. For example: text, rich text, buttons, images, or a spinner.
> 1. `footer-actions` slot element which is used to add actions buttons after the main content.
> 1. `followup-prompt` slot element which is used to add followup prompt buttons. They are below any action buttons.

Reviewer: Small miscommunication from my previous comment, we prefer "slot" over "slot element". Also changing the language to be more precise, since "after" could mean "below" or "right of"

```suggestion
1. `default` slot displays arbitrary slotted content. For example: text, rich text, buttons, images, or a spinner.
1. `footer-actions` slot which is used to add actions buttons below the main content.
1. `followup-prompt` slot which is used to add followup prompt buttons. They are below any action buttons.
```

Author: Made the suggested edits.

> 1. `followup-prompt` slot element which is used to add followup prompt buttons. They are below any action buttons.

Reviewer: Now that we're settled on the follow up prompt being in a slot rather than a separate message, we should agree on the name. I'd like to avoid "prompt" since it's AI-specific. Alternatives:

1. `slim-footer-actions` and `footer-actions`
2. `footer-actions` and `footer-2-actions`
3. `footer-actions` and `sub-footer-actions`

At the moment I like 1 best but I'll ping other owners for their opinions.

Reviewer 2: For implementing the `end` slot we should follow the exisiting start-end pattern.

So for example, see how the combobox applies the `StartEnd` mixin to the class:
https://github.com/ni/nimble/blob/175f1ab564314a51cb6eb6446dc1eb5eff3805f1/packages/nimble-components/src/combobox/index.ts#L885-L886

It then uses the `startSlotTemplate` and `endSlotTemplate` helpers in the template:
https://github.com/ni/nimble/blob/175f1ab564314a51cb6eb6446dc1eb5eff3805f1/packages/nimble-components/src/combobox/template.ts#L40
https://github.com/ni/nimble/blob/175f1ab564314a51cb6eb6446dc1eb5eff3805f1/packages/nimble-components/src/combobox/template.ts#L69

And then you can style the slot using the part selector `[part="end"] { /* styles */ }`.

Currently we are only using the `end` template but we should still include both start and end, for example:

```js
    <div class="root">
        ${startSlotTemplate(context, definition)}
        <section class="message-content">
            <slot></slot>
        </section>
        <section>
           <slot class="footer-actions" name="footer-actions"></slot>
        </section>
        ${endSlotTemplate(context, definition)}
    </div>
```

But we should hide the start template with styles since it is unneeded currently:

```css
        [part='start'] {
            display: none;
        }

        [part='end'] {
             /* Styles */
        }

```

Author: Changed slot name to `end` and started using the start-end pattern.

> All action buttons must meet the following criteria

Reviewer: We should document similar criteria for the followup slot. I suspect we want them all to be block buttons with text rather than icons, but that's UX's call.

Author: Added criteria for the end slot.  For now, I added that the appearance for a button should be `Outline` since that was what I was in many of the Figma explorations.  I asked Brandon for an official design for the action and follow up buttons.

I plan to submit the changes that I have now since I don't know how long it will take Brandon to give me the design. I will submit any updates on a different PR.

> 1. They are `nimble-button`s

Author: Added that `nimble-toggle-buttons` can also be used.

> 1. They are `nimble-button`s

Reviewer 3: If that's the case, then doesn't the styling need to be updated to support toggle buttons? As is, it only supports `buttonTag`.

> 1. They are `nimble-button`s

Author: Update styling to include toggle buttons.

#### Chat message

> The `spright-chat-message` is divided into four sections.

Author: Updated description to use "slot elements".  I will update the names after our meeting.

> 

Author: We met offline and discuss this feedback.  This was the outcome
1. Removed the left slot.  We cannot keyboard navigate to the hidden edit button and we did not want to always show the button since it would add noise.  So, we decided to not add the edit button on the left slot for now which meant also removing the left slot.
2. We decided to renamed the left-bottom slot to footer-actions.
3. We decided to keep the followup-prompt slot and the name.
4. I tried using a nimble-toolbar for the footer-action slot, but the toolbar did not support adding slotted elements to the toolbar.  An issue will be created to support this use case.

#### Chat message

The `spright-chat-message` has the following slot elements.

1. `default` slot displays arbitrary slotted content. For example: text, rich text, buttons, images, or a spinner.

> 1. `footer-actions` slot which is used to add actions buttons below the main content.

Reviewer: ```suggestion
1. `footer-actions` slot which is used to add action buttons below the main content.
```

Author: Updated

1. `end` slot which is used to add text buttons. They are below any action buttons.

> All action buttons must meet the following criteria
> 
> 1. They are `nimble-button`s or `nimble-toggle-button`
> 1. The `appearance` attribute is set to `ghost`
> 1. The `content-hidden` attribute is set to `true`
> 1. They contain only Nimble icons
> 
> All end text buttons must meet the following criteria
> 
> 1. They are `nimble-button`s
> 1. The `apperance` attribute is set to `block`
> 1. The buttons only have text

Reviewer: nit: copy edits
```suggestion
Nimble will set the height of the action buttons to `$ni-nimble-control-slim-height`. All action buttons must meet the following criteria:

1. They are `nimble-button`s or `nimble-toggle-button`s
1. The `appearance` attribute is set to `ghost`
1. The `content-hidden` attribute is set to `true`
1. They contain only Nimble icons

All end text buttons must meet the following criteria:

1. They are `nimble-button`s
1. The `apperance` attribute is set to `block`
1. The buttons only have text
```

Author: Added new text

    - A message will grow its height to fit its content, with no maximum height.
    - Clients could override this behavior but we don't anticipate use cases for doing so when the message is used within a conversation
- _Slots_
    - `footer-actions`
        - Action buttons to display after the main content.
    - `end`

>         - Buttons with text that are display at the bottom after any action buttons.

Reviewer: ```suggestion
        - Buttons with text that are displayed at the bottom after any action buttons.
```

Author: Updated

### Blazor integration

> The Blazor wrappers `SpringChatConversation` and `SprightChateMessage` have been created.

Reviewer: ```suggestion
The Blazor wrappers `SpringChatConversation` and `SprightChatMessage` have been created.
```

Author: Fixed mispelling

### Visual Appearance

> 

Reviewer: Let's mention the toolbar alternative and link to the issue from the spec somewhere. You could do that in the Accessibility section or an Alternatives section.

Author: Added to the Accessibility section and added a link to the toolbar issue.

### Blazor integration

The Blazor wrappers `SpringChatConversation` and `SprightChateMessage` have been created.

### Visual Appearance

> 

Reviewer: If we don't end up implementing the keyboard-accessible pattern for interacting with the hover-only edit button, we should start keeping a list of known accessibility gaps in the Accessibility section below.

Author: Since we no longer have an edit button, there is nothing to do so resolving this comment.

### Accessibility

> Only keyboard navigation has been evaluated. This has led to the removal of a showing an action button only on hover. Users cannot navigate to a hidden button. We wil revisit this in the future if desired behavior is required.
> 
> For the footer actions elements, they need to be in a `nimble-toolbar`, however, the `nimble-toolbar` does not support content slotted across a shadow root. [This](https://github.com/ni/nimble/issues/2571) issue has been opened to fix it.

Reviewer: ```suggestion
Only keyboard navigation has been evaluated. The desired behavior is for each message's action buttons to be a single tab stop, with navigation between a message's action buttons accomplished using arrow keys. This should be achieved by placing the items within a `nimble-toolbar`, however the `nimble-toolbar` does not yet detect content in nested slots (see #2571). For scoping reasons the content will not initially be placed in a toolbar and thus each button will be a tab stop. Once that issue is addressed it should be possible to add a toolbar and achieve the desired behavior without breaking clients.

Buttons placed in the `end` slot should each be their own tab stop and thus will not be placed in a toolbar.

The design spec includes a proposal to add another "Edit" button to the left of some messages when the user hovers over the message. The button would be a tab stop and would become visible when keyboard focused. This work is currently out of scope.

Other aspects of accessibility have not yet been evaluated.
```

Author: Updated text as suggested.
